rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,243 @@
+| Conv/IP/Matmul            | attr::output_scale          |
+|                           | attr::p_op_binary_mul       |
+| RNN                       | attr::rnn_data_qparams      |
+|                           | attr::rnn_weights_qparams   |
+| Batch/Layer normalization |                             |
+| Binary                    | attr::scales                |
+| attr::p_op::binary        |                             |
+| Concat                    |                             |
+| Eltwise                   |                             |
+| attr::p_op::eltwise       | built-in parameter          |
+| LRN                       |                             |
+| (Log)Softmax              | attr::output_scale[[#2]][2] |
+| Pooling                   | attr::p_op::binary_mul      |
+| PRelu                     |                             |
+| Resampling                | attr::p_op::binary_mul      |
+| Shuffle                   |                             |
+| Sum                       |                             |
+| attr::p_op::sum           | built-in parameter          |
+| Reorder                   | attr::output_scale          |
+| Reduction                 | attr::p_op::binary_mul      |
+
+
+## 2. Addressing short-term requests
+
+### 2.1. Run-time scaling in Eltwise
+
+This feature doesn't require API extensions, because scaling can be implemented
+as a separate Binary post-op.
+
+
+### 2.2. Scaling for src1 in Binary post-op
+
+New API is required to support the requested feature. New API extends binary
+post-op with a `scale` parameter (scaling mechanism `built-in parameter`). This
+mechanism is aligned with scaling for other post-operations and shares the same
+limitations:
+- only single-value scaling;
+- no run-time support.
+
+API:
+
+```c
+dnnl_status_t DNNL_API dnnl_post_ops_append_binary_v2(dnnl_post_ops_t post_ops,
+        float scale, dnnl_alg_kind_t alg_kind,
+        const dnnl_memory_desc_t *src1_desc);
+
+dnnl_status_t DNNL_API dnnl_post_ops_get_params_binary_v2(
+        const_dnnl_post_ops_t post_ops, int index, float *scale,
+        dnnl_alg_kind_t *alg_kind, const dnnl_memory_desc_t **src1_desc);
+```
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(scale, algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+```
+
+An alternative solution would be to extend mechanism `attr::scales`. This
+solution doesn't require new API, but extends the existing one.
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+```
+
+While the alternative solution allows to avoid new API, it introduces one more
+way to set scale which complicates both implementation and integration.
```

Comment:
Could you develop why relying on `set_scales` would be harder in implementation and integration?

Also, I agree that the `dnnl_post_ops_append_binary_v2` makes it "more consistent" with other post_ops we currently have. However, introducing `dnnl_post_ops_append_binary_v2` does not seem to align with our long term goal.  IMHO the biggest consideration here should be how soon we want the long term plan to realize, and how we want to proceed. We can either:
- introduce a new symbol to try to keep some level of alignment, even though it means deprecating it for set_scales in the future. The first option would go in that direction.
- or have new features align on the long term scheme (set_scales), that we would not deprecate, even if it means less alignment in short term. The second option would go in that direction

I would vote for the later, since we don't actually have alignment currently (we kind of have a zoo as you outlined in previous section), and it would minimize future changes on the user side.

Reply:
Sounds good to me. We can use `set_scales` in binary post op as a testing of the long-term solution.

Initially I looked at these 2 options as equal, but I decided to go with `dnnl_post_ops_append_binary_v2` only to avoid different mechanisms for scaling in post-ops -- On an application side the second option will introduces code branches.